src/MicroHs/Lex.hs

```diff
@@ -80,10 +81,24 @@ lex loc (d:cs) | isLower_ d =
   case span isIdentChar cs of
     (ds, rs) -> tIdent loc [] (d:ds) (lex (addCol loc $ 1 + length ds) rs)
 lex loc cs@(d:_) | isUpper d = upperIdent loc loc [] cs
-lex loc ('0':x:cs) | toLower x == 'x' = hexNumber loc cs
-                   | toLower x == 'o' = octNumber loc cs
-                   | toLower x == 'b' = binNumber loc cs
-lex loc cs@(d:_) | isDigit d = number loc cs
+lex loc ('0':x:cs)
+  | toLower x == 'x' =
```

Comment:
There's a lot of code duplication here.  You can make a new function alleviate that.